numba_cuda/numba/cuda/cudadrv/managed_module.py

```diff
@@ -0,0 +1,93 @@
+import weakref
+
+from . import devices
+from .driver import CtypesModule, USE_NV_BINDING
+
+
+class _CuFuncProxy:
+    """See `ManagedModule` for more details
+    """
+    def __init__(self, module, cufunc):
+        self._module = module
+        self._cufunc = cufunc
+
+    def init_module(self, stream):
+        self._module._init(stream)
+
+    def lazy_finalize_module(self, stream):
+        self._module._lazy_finalize(stream)
+
+    def __getattr__(self, name):
+        return getattr(self._cufunc, name)
+
+
+class ManagedModule:
+    def __init__(self, module, setup_callbacks, teardown_callbacks):
+        """ctypes module with setup and teardown callbacks
+        The use of managedmodule is the same as a ctypes module,
+        with the exception of `get_function`, which returns a wrapped
+        cufunc object. The wrapped object provides `init_module`
+        and `lazy_finalize_module` method. Which are used
+        to initialize and finalize the module when stream is available
+        in the later stage of the compilation pipeline.
+        """
+        # To be updated to object code
+        if not isinstance(module, CtypesModule):
+            raise TypeError("module must be a CtypesModule")
+
+        if not isinstance(setup_callbacks, list):
+            raise TypeError("setup_callbacks must be a list")
+        if not isinstance(teardown_callbacks, list):
+            raise TypeError("teardown_callbacks must be a list")
+
+        for callback in setup_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+        for callback in teardown_callbacks:
+            if not callable(callback):
+                raise TypeError("all callbacks must be callable")
+
+        self._initialized = False
+        self._module = module
+        self._setup_callbacks = setup_callbacks
+        self._teardown_callbacks = teardown_callbacks
+
+    def _init(self, stream):
+        """Eagerly call the setup functions for cumodule on `stream`
+        """
+        for setup in self._setup_callbacks:
+            setup(self._module, stream)
+
+    def _lazy_finalize(self, stream):
+        """Set teardown function for cumodule via weakref finalizer.
+        """
+        def lazy_callback(callbacks, module, stream):
+            for teardown in callbacks:
+                teardown(module, stream)
+
+        ctx = devices.get_context()
+        if USE_NV_BINDING:
+            key = self._module.handle
+        else:
+            key = self._module.handle.value
+        module_obj = ctx.modules.get(key, None)
+
+        if module_obj is not None:
```

Comment:
Should there be an `else` branch that raises an exception if we can't find the module? Are there any circumstances in which we could attempt to set a finalizer for a non-existent module?

Reply:
I'm really not sure. I could perform a print to see if in any scenario within the test suite the module would be non-existent.

Reply:
I think I'd err on the side of caution, and raise an exception if the module can't be found - if we didn't find one, I think we're going to omit a call to a teardown function that I think would otherwise have been expected to happen.

Reply:
After refactoring the code into `Module`, this issue naturally goes away because `self` is an owning object of the module. We no longer need to retrieve it from the context.